File: app/src/test/java/it/niedermann/owncloud/notes/persistence/WidgetNotesListDaoTest.java
/*
 * Nextcloud Notes - Android Client
 *
 * SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
package it.niedermann.owncloud.notes.persistence;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import androidx.annotation.NonNull;
import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.room.Room;
import androidx.test.core.app.ApplicationProvider;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import it.niedermann.owncloud.notes.persistence.entity.Account;
import it.niedermann.owncloud.notes.persistence.entity.NotesListWidgetData;
import it.niedermann.owncloud.notes.shared.model.Capabilities;

@RunWith(RobolectricTestRunner.class)
public class WidgetNotesListDaoTest {

    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    @NonNull
    private NotesDatabase db;

    private long accountId;

    @Before
    public void setupDB() {
        db = Room
                .inMemoryDatabaseBuilder(ApplicationProvider.getApplicationContext(), NotesDatabase.class)
                .allowMainThreadQueries()
                .build();
        accountId = db.getAccountDao().insert(new Account("https://example.com", "user", "user@example.com", null, new Capabilities()));
    }

    @After
    public void closeDb() {
        db.close();
    }

    @Test
    public void insertThenReadWidgetData() {
        final var data = widgetData(1, NotesListWidgetData.MODE_DISPLAY_ALL);
        db.getWidgetNotesListDao().createOrUpdateNoteListWidgetData(data);

        final var loaded = db.getWidgetNotesListDao().getNoteListWidgetData(1);
        assertNotNull(loaded);
        assertEquals(NotesListWidgetData.MODE_DISPLAY_ALL, loaded.getMode());
    }

    /**
     * Reconfiguring an interactive widget saves again for the same appWidgetId (primary key).
     * The DAO must upsert; a plain insert would abort on the primary-key conflict.
     */
    @Test
    public void reconfigureUpdatesExistingWidgetData() {
        db.getWidgetNotesListDao().createOrUpdateNoteListWidgetData(widgetData(1, NotesListWidgetData.MODE_DISPLAY_ALL));

        db.getWidgetNotesListDao().createOrUpdateNoteListWidgetData(widgetData(1, NotesListWidgetData.MODE_DISPLAY_STARRED));

        final var loaded = db.getWidgetNotesListDao().getNoteListWidgetData(1);
        assertNotNull(loaded);
        assertEquals(NotesListWidgetData.MODE_DISPLAY_STARRED, loaded.getMode());
    }

    @NonNull
    private NotesListWidgetData widgetData(int appWidgetId, int mode) {
        final var data = new NotesListWidgetData();
        data.setId(appWidgetId);
        data.setAccountId(accountId);
        data.setMode(mode);
        data.setCategory(null);
        return data;
    }
}

File: app/src/test/java/it/niedermann/owncloud/notes/persistence/WidgetNotesListDaoTest.kt
/*
 * Nextcloud Notes - Android Client
 *
 * SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
package it.niedermann.owncloud.notes.persistence

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import it.niedermann.owncloud.notes.persistence.entity.Account
import it.niedermann.owncloud.notes.persistence.entity.NotesListWidgetData
import it.niedermann.owncloud.notes.shared.model.Capabilities
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [36])
class WidgetNotesListDaoTest {

    @get:Rule
    val instantTaskExecutorRule = InstantTaskExecutorRule()

    private lateinit var db: NotesDatabase
    private var accountId: Long = 0

    @Before
    fun setupDB() {
        db = Room
            .inMemoryDatabaseBuilder(ApplicationProvider.getApplicationContext(), NotesDatabase::class.java)
            .allowMainThreadQueries()
            .build()

        val account = Account(ACCOUNT_URL, ACCOUNT_USERNAME, ACCOUNT_NAME, null, Capabilities())
        accountId = db.accountDao.insert(account)
    }

    @After
    fun closeDb() {
        db.close()
    }

    @Test
    fun insertThenReadWidgetData() {
        db.widgetNotesListDao.createOrUpdateNoteListWidgetData(
            widgetData(NotesListWidgetData.MODE_DISPLAY_ALL)
        )

        val loaded = db.widgetNotesListDao.getNoteListWidgetData(WIDGET_ID)
        assertNotNull(loaded)
        assertEquals(NotesListWidgetData.MODE_DISPLAY_ALL, loaded.mode)
    }

    /**
     * Reconfiguring an interactive widget saves again for the same appWidgetId (primary key).
     * The DAO must upsert; a plain insert would abort on the primary-key conflict.
     */
    @Test
    fun reconfigureUpdatesExistingWidgetData() {
        db.widgetNotesListDao.createOrUpdateNoteListWidgetData(
            widgetData(NotesListWidgetData.MODE_DISPLAY_ALL)
        )
        db.widgetNotesListDao.createOrUpdateNoteListWidgetData(
            widgetData(NotesListWidgetData.MODE_DISPLAY_STARRED)
        )

        val loaded = db.widgetNotesListDao.getNoteListWidgetData(WIDGET_ID)
        assertNotNull(loaded)
        assertEquals(NotesListWidgetData.MODE_DISPLAY_STARRED, loaded.mode)
    }

    private fun widgetData(mode: Int) = NotesListWidgetData().apply {
        id = WIDGET_ID
        accountId = this@WidgetNotesListDaoTest.accountId
        this.mode = mode
        category = null
    }

    companion object {
        private const val WIDGET_ID = 1
        private const val ACCOUNT_URL = "https://example.com"
        private const val ACCOUNT_USERNAME = "user"
        private const val ACCOUNT_NAME = "user@example.com"
    }
}
